### ISSUE TITLE
2FA with Okta failing

### ISSUE THREAD

**Issue by bt-** [yuezk/GlobalProtect-openconnect#152]
Hi, this application has been working great for me and I was very relieved to find it after switching from OSX to linux.

One of the VPNs that I have to connect to just switched to using Okta 2FA. It was working fine before they integrated Okta, but now it is not working. I can login and the 2FA prompt shows up without any issues. I approve the push notification and then it fails. I've pasted the log below. Any suggestions would be very much appreciated.

```
2022-05-18 17:34:45.822 INFO  [133509] [main@25] GlobalProtect started, version: 1.3.4
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
2022-05-18 17:34:46.024 INFO  [133509] [GPClient::populateGatewayMenu@140] Populating the Switch Gateway menu...
2022-05-18 17:34:49.137 INFO  [133509] [GPClient::populateGatewayMenu@140] Populating the Switch Gateway menu...
2022-05-18 17:34:53.314 INFO  [133509] [GPClient::doConnect@245] Start connecting...
2022-05-18 17:34:53.314 INFO  [133509] [GPClient::doConnect@266] Start portal login...
2022-05-18 17:34:53.316 INFO  [133509] [PortalAuthenticator::authenticate@33] Preform portal prelogin at https://X.X.XXX.XX/global-protect/prelogin.esp?tmp=tmp&kerberos-support=yes&ipv6-support=yes&clientVer=4100&clientos=Linux
2022-05-18 17:34:53.579 INFO  [133509] [PortalAuthenticator::onPreloginFinished@50] Portal prelogin succeeded.
2022-05-18 17:34:53.579 INFO  [133509] [PreloginResponse::parse@26] Start parsing the prelogin response...
2022-05-18 17:34:53.579 INFO  [133509] [PortalAuthenticator::onPreloginFinished@54] Finished parsing the prelogin response. The region field is: US
2022-05-18 17:34:53.579 INFO  [133509] [PortalAuthenticator::samlAuth@121] Trying to perform SAML login with saml-method POST

DevTools listening on ws://127.0.0.1:12315/devtools/browser/935d8261-f86c-41fb-8a1f-61b46c57a33f
Remote debugging server started successfully. Try pointing a Chromium-based browser to http://127.0.0.1:12315
2022-05-18 17:34:53.733 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from data:text/html;charset=UTF-8,%3Chtml%3E%0A%3Cbody%3E%0A%3Cform%20id%3D%22myform%22%20method%3D%22POST%22%20action%3D%22https%3A%2F%2Fgridsec.okta.com%2Fapp%2Fpanw_globalprotect%2Fexkg9uh21vGE9emlW2p7%2Fsso%2Fsaml%22%3E%0A%3Cinput%20type%3D%22hidden%22%20name%3D%22SAMLRequest%22%20value%3D%22PHNhbWxwOkF1dGhuUmVxdWVzdCB4bWxuczpzYW1scD0idXJuOm9hc2lzOm5hbWVzOnRjOlNBTUw6Mi4wOnByb3RvY29sIiBBc3NlcnRpb25Db25zdW1lclNlcnZpY2VVUkw9Imh0dHBzOi8vOC43LjE1My42Nzo0NDMvU0FNTDIwL1NQL0FDUyIgRGVzdGluYXRpb249Imh0dHBzOi8vZ3JpZHNlYy5va3RhLmNvbS9hcHAvcGFud19nbG9iYWxwcm90ZWN0L2V4a2c5dWgyMXZHRTllbWxXMnA3L3Nzby9zYW1sIiBJRD0iXzBhZDRkNzEyMDY5MTUxNGQ4YTIyZmQ1Yzk5ZjI4ZTVhIiBJc3N1ZUluc3RhbnQ9IjIwMjItMDUtMThUMjI6MzQ6NTNaIiBQcm90b2NvbEJpbmRpbmc9InVybjpvYXNpczpuYW1lczp0YzpTQU1MOjIuMDpiaW5kaW5nczpIVFRQLVBPU1QiIFZlcnNpb249IjIuMCI%2BPHNhbWw6SXNzdWVyIHhtbG5zOnNhbWw9InVybjpvYXNpczpuYW1lczp0YzpTQU1MOjIuMDphc3NlcnRpb24iPmh0dHBzOi8vOC43LjE1My42Nzo0NDMvU0FNTDIwL1NQPC9zYW1sOklzc3Vlcj48L3NhbWxwOkF1dGhuUmVxdWVzdD4%3D%22%20%2F%3E%0A%3Cinput%20type%3D%22hidden%22%20name%3D%22RelayState%22%20value%3D%22YVeWLgAALRM1MWQxZjgwZjA1NjI2Y2JkMjE0NWEzMTIxZDhlZWZhMQ%3D%3D%22%20%2F%3E%0A%3C%2Fform%3E%0A%3Cscript%3E%0A%20%20document.getElementById%28%27myform%27%29.submit%28%29%3B%0A%3C%2Fscript%3E%0A%3C%2Fbody%3E%0A%3C%2Fhtml%3E%0D%0A
2022-05-18 17:34:53.839 INFO  [133509] [SAMLLoginWindow::onLoadFinished@98] Load finished https://X.X.XXX.XX/global-protect/prelogin.esp?tmp=tmp&kerberos-support=yes&ipv6-support=yes&clientVer=4100&clientos=Linux
2022-05-18 17:34:54.232 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://gridsec.okta.com/app/panw_globalprotect/exkg9uh21vGE9emlW2p7/sso/saml
2022-05-18 17:34:54.479 INFO  [133509] [SAMLLoginWindow::onLoadFinished@98] Load finished https://gridsec.okta.com/app/panw_globalprotect/exkg9uh21vGE9emlW2p7/sso/saml
2022-05-18 17:34:54.492 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://login.okta.com/discovery/iframe.html
2022-05-18 17:35:05.118 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://gridsec.okta.com/auth/services/devicefingerprint
2022-05-18 17:35:05.790 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://gridsec.okta.com/auth/services/devicefingerprint
2022-05-18 17:35:08.003 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://gridsec.okta.com/auth/services/devicefingerprint
2022-05-18 17:35:16.177 INFO  [133509] [SAMLLoginWindow::onResponseReceived@64] Response received from https://gridsec.okta.com/login/sessionCookieRedirect
2022-05-18 17:35:16.221 INFO  [133509] [SAMLLoginWindow::onLoadFinished@98] Load finished https://gridsec.okta.com/login/sessionCookieRedirect
[133662:133682:0518/173516.306240:ERROR:ssl_client_socket_impl.cc(960)] handshake failed; returned -1, SSL error code 1, net_error -202
2022-05-18 17:35:16.337 INFO  [133509] [SAMLLoginWindow::onLoadFinished@98] Load finished https://X.X.X.XX/SAML20/SP/ACS
```

**Comment by bt-** [comment]
I was able to login from a Windows virtual machine, which is now my backup plan. When I logged in from Windows using the Global Protect client, after approve the 2FA push notification there was a second certificate warning that I had to accept to complete the connection. I am already supplying a custom parameter for `--servercert` to get as far as I had. Maybe there is a way to supply another one?

**Comment by techgaun** [comment]
I am also running into this same issue. Pretty much the same log:
```
2022-06-01 11:19:35.472 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://<private-tenant-stripped>.okta.com/app/panw_globalprotect/exkiav15pQrXTXUvT696/sso/saml
2022-06-01 11:19:35.811 INFO  [107681] [SAMLLoginWindow::onLoadFinished@99] Load finished https://<private-tenant-stripped>.okta.com/app/panw_globalprotect/exkiav15pQrXTXUvT696/sso/saml
2022-06-01 11:19:35.827 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://login.okta.com/discovery/iframe.html
2022-06-01 11:19:52.442 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://<private-tenant-stripped>.okta.com/auth/services/devicefingerprint
2022-06-01 11:19:54.602 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://<private-tenant-stripped>.okta.com/login/sessionCookieRedirect
2022-06-01 11:19:54.713 INFO  [107681] [SAMLLoginWindow::onLoadFinished@99] Load finished https://<private-tenant-stripped>.okta.com/login/sessionCookieRedirect
2022-06-01 11:19:54.723 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://login.okta.com/discovery/iframe.html
2022-06-01 11:20:06.478 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://<private-tenant-stripped>.okta.com/login/sessionCookieRedirect
2022-06-01 11:20:06.534 INFO  [107681] [SAMLLoginWindow::onLoadFinished@99] Load finished https://<private-tenant-stripped>.okta.com/login/sessionCookieRedirect
2022-06-01 11:20:07.945 INFO  [107681] [SAMLLoginWindow::onResponseReceived@65] Response received from https://gw.paylease.com/SAML20/SP/ACS/SAML20/SP/ACS
2022-06-01 11:20:07.966 INFO  [107681] [SAMLLoginWindow::onLoadFinished@99] Load finished https://gw.paylease.com/SAML20/SP/ACS/SAML20/SP/ACS
```

The final page says "Login Successful" but the authentication does not seem to complete. @yuezk would have any idea or is there anyway I can debug this further?

**Comment by EmptyVisual** [comment]
Reiterating the above:

Upon submitting the 6 digit code successfully for Okta, The main window displays 'Authenticating' while the Okta/Auth window says 'Login Successful', nothing moves beyond that point.

Screenshot:

![image](https://user-images.githubusercontent.com/4710051/171770866-ef043211-cfdb-4797-903f-821dc10fe6d2.png)

**Comment by techgaun** [comment]
Ok I think I was able to debug this further little more. The last URL (`https://<private-url>/SAML20/SP/ACS/SAML20/SP/ACS`) when loaded has the necessary content as html comment.
```
<html>
<body>Login Successful!</body>
<!--<saml-auth-status>1</saml-auth-status><prelogin-cookie>PRELOGIN_COOKIE_HERE</prelogin-cookie><saml-username>USER_EMAIL</saml-username><saml-slo>no</saml-slo><saml-SessionNotOnOrAfter></saml-SessionNotOnOrAfter>-->
</html>
```

So this needs to be handled so that those parameters are extracted from the html body comment instead of the headers.

**Comment by techgaun** [comment]
ok I've created this PR: https://github.com/yuezk/GlobalProtect-openconnect/pull/156
The code in that PR works fine for me. Hopefully others can try and also improve the code based on the things I was able to discover. @bt- and @EmptyVisual maybe you guys can give a try the PR I've created ([branch here](https://github.com/techgaun/GlobalProtect-openconnect/tree/okta-2fa-fix))

**Comment by EmptyVisual** [comment]
@techgaun PR #156 works and and connects successfully with the main window showing as Connected. Thank you!

![image](https://user-images.githubusercontent.com/4710051/172052217-cd6f27c8-fc9c-424a-8d59-ee52a95e2a44.png)

**Comment by yuezk** [comment]
@techgaun Thanks for your PR. I will review and merge it ASAP.

**Comment by techgaun** [comment]
awesome thanks @yuezk any plans on creating new release?

**Comment by bt-** [comment]
I couldn't test this until now because I needed to update my OS to update this Globalprotect-openconnect. I did that this week and was just trying to connect to a VPN with 2FA through Okta and so far I'm not able to get it to work. I'm using 1.4.8 and I added the servercert to the config file. I was able to connect to a site that is not using 2FA.
I was also able to connect to the site that is failing on Linux through my windows VM using the regular GlobalProtect client.
I was trying to collect the logs, but now I'm running into a different issue maybe from trying to login to many times.
